streams/src/test/java/org/apache/kafka/streams/integration/KStreamRepartitionIntegrationTest.java

```diff
@@ -0,0 +1,548 @@
+import org.apache.kafka.test.IntegrationTest;
+import org.apache.kafka.test.TestUtils;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.Properties;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+@Category({IntegrationTest.class})
+public class KStreamRepartitionIntegrationTest {
+    private static final int NUM_BROKERS = 1;
+    private static AtomicInteger testNo = new AtomicInteger(0);
+
+    @ClassRule
+    public static final EmbeddedKafkaCluster CLUSTER = new EmbeddedKafkaCluster(NUM_BROKERS);
+
+    private Properties streamsConfiguration;
+    private List<KafkaStreams> kafkaStreamsInstances;
+    private String inputTopic = "input-topic-" + testNo.get();
+    private String outputTopic = "output-topic-" + testNo.get();
+    private String applicationId = "kstream-repartition-stream-test-" + testNo.get();
+
+    @Before
+    public void before() throws InterruptedException {
+        CLUSTER.createTopic(inputTopic, 4, 1);
+        CLUSTER.createTopic(outputTopic, 1, 1);
+
+        streamsConfiguration = new Properties();
+        kafkaStreamsInstances = new ArrayList<>();
+
+        streamsConfiguration.put(StreamsConfig.APPLICATION_ID_CONFIG, applicationId);
+        streamsConfiguration.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, CLUSTER.bootstrapServers());
+        streamsConfiguration.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+        streamsConfiguration.put(StreamsConfig.STATE_DIR_CONFIG, TestUtils.tempDirectory().getPath());
+        streamsConfiguration.put(StreamsConfig.CACHE_MAX_BYTES_BUFFERING_CONFIG, 0);
+        streamsConfiguration.put(StreamsConfig.COMMIT_INTERVAL_MS_CONFIG, 100);
+        streamsConfiguration.put(StreamsConfig.DEFAULT_KEY_SERDE_CLASS_CONFIG, Serdes.Integer().getClass());
+        streamsConfiguration.put(StreamsConfig.DEFAULT_VALUE_SERDE_CLASS_CONFIG, Serdes.String().getClass());
+        streamsConfiguration.put(StreamsConfig.TOPOLOGY_OPTIMIZATION, StreamsConfig.OPTIMIZE);
+    }
+
+    @After
+    public void whenShuttingDown() throws IOException {
+        kafkaStreamsInstances
+            .stream()
+            .filter(Objects::nonNull)
+            .forEach(KafkaStreams::close);
+
+        IntegrationTestUtils.purgeLocalStreamsState(streamsConfiguration);
+        testNo.incrementAndGet();
+    }
+
+    @Test
+    public void shouldNotOptimizeRepartitionOperationTopology() throws ExecutionException, InterruptedException {
```

Comment:
This can (and should) be a unit test, since we don't need to produce data or run Kafka to build and verify the topology.

Reply:
Added `StreamsGraphTest#shouldNotOptimizeWhenRepartitionOperationIsDone`